dotnet/src/InternalUtilities/src/Data/VectorStoreRecordPropertyReader.cs

```diff
@@ -337,6 +338,61 @@ public string GetJsonPropertyName(string dataModelPropertyName)
         return this._jsonPropertyNamesMap.Value[dataModelPropertyName];
     }
 
+    /// <summary>
+    /// Get the <see cref="VectorStoreRecordVectorProperty"/> for the given property name.
+    /// </summary>
+    /// <exception cref="InvalidOperationException">When there is no match.</exception>
+    public VectorStoreRecordVectorProperty GetVectorProperty<TRecord>(VectorSearchOptions<TRecord>? searchOptions)
+    {
+        if (searchOptions is not null)
+        {
+#pragma warning disable CS0618 // Type or member is obsolete
+            string? propertyName = searchOptions.VectorPropertyName;
+#pragma warning restore CS0618 // Type or member is obsolete
+
+            if (searchOptions.VectorProperty is Expression<Func<TRecord, object?>> expression)
+            {
+                if (expression.Body is ConstantExpression constant && constant.Value is string text && !string.IsNullOrWhiteSpace(text))
+                {
+                    // This is something that VectorStoreGenericDataModel can use.
+                    propertyName = text;
+                }
+                // r => r.PropertyName is translated into
+                // r => (object)r.PropertyName
+                else if (expression.Body is UnaryExpression unary
+                    && unary.Operand.NodeType == ExpressionType.MemberAccess
+                    && unary.Operand is MemberExpression member
+                    && expression.Parameters.Count == 1
+                    && member.Expression == expression.Parameters[0]
+                    && member.Member is PropertyInfo property)
```

Comment:
C# pattern matching works especially well in cases like this:

```suggestion
                else if (expression is
                         {
                             Body: UnaryExpression
                             {
                                 NodeType: ExpressionType.Quote,
                                 Operand: MemberExpression member
                             },
                             Parameters: [var lambdaParameter]
                         }
                         && member.Expression == lambdaParameter
                         && member.Member is PropertyInfo property)
```

Reply:
I am sorry, but I don't like this syntax ;) I've spent too much time writing C for runtime ;)